### PR TITLE
[RfR] Registration links; Serializer tests

### PR DESCRIPTION
**Pull request by sloria**
### Purpose/Changes

- Add initial unit tests for serializers 
- Add a few missing links and attributes to NodeSerializer and NodeRegistrationSerializer (partially addresses OSF-4436)
- Change NodeSerializer.parent to a `related` link rather than a `self` link (verified with Dawn that this is correct).

### Tickets

https://openscience.atlassian.net/browse/OSF-4771
Partially addresses https://openscience.atlassian.net/browse/OSF-4436

### CODE REVIEW

File: tests/api_tests/nodes/test_serializers.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+from tests.base import DbTestCase, ApiTestCase, assert_datetime_equal
+from tests.utils import make_drf_request
+from tests.factories import UserFactory, NodeFactory, RegistrationFactory, ProjectFactory
+
+from framework.auth import Auth
+from api.nodes.serializers import NodeSerializer, NodeRegistrationSerializer
+from api.base.settings.defaults import API_BASE
+
+
+class TestNodeSerializer(DbTestCase):
```
Comment (brianjgeiger) [inline tests/api_tests/nodes/test_serializers.py]: I'm not opposed to the serializer tests, but I'm curious how they fit in to the rest of the tests. Essentially just making sure that nobody removes or renames things (which is good)?

File: website/project/model.py [hunk 2]
```diff
@@ -730,6 +731,23 @@ def category_display(self):
         """The human-readable representation of this node's category."""
         return self.CATEGORY_MAP[self.category]
 
+    # We need the following 2 properties in order to serialize related links in NodeRegistrationSerializer
+    @property
+    def registered_user_id(self):
```
Comment (brianjgeiger) [inline website/project/model.py]: Look at you: changing the model to suit the API. You've gone mad with power.

**Comment by brianjgeiger** [comment]
The self link does not take you to a representation that matches the view in nodes/registrations. So if you want to have a registration serializer, we probably need to make a collection for registrations that will also give the proper detail view. We could, potentially, just roll all that up into the node serializer, but that will get messy with nodes.

Also, if we have a collection for registrations, we'll probably not want to include them in the nodes collection.

File: website/project/model.py (hunk 2 above)
Comment (sloria) [inline website/project/model.py]: Yes, I really don't like having to do this, but DRF won't let us do `lookup_field_name='registered_user._id'; it only allows simple attribute access. If you have suggestions for an immediate fix, I am open to them.

FWIW, I do intend to address this limitation if we go with my proposed `RelationshipField`.

**Comment by brianjgeiger** [comment]
Popping this out of CR1 while we consider philosophical questions and make changes to that. Will continue once resolved.

File: website/project/model.py (hunk 2 above)
Comment (brianjgeiger) [inline website/project/model.py]: Traditionally we handle that in the view, but that, also, is unsatisfying for different reasons. This, at least, keeps us from having to iterate over the entire list of results each time.

File: tests/api_tests/nodes/test_serializers.py (hunk 1 above)
Comment (sloria) [inline tests/api_tests/nodes/test_serializers.py]: Correct; they ensure that we don't break API contracts w.r.t. response format without having to do a high-level test like a views test or Runscope test.

**Comment by sloria** [comment]
> The self link does not take you to a representation that matches the view in nodes/registrations.

Can you clarify which `self` link you are referring to? I didn't think I added a self link in this PR.

**Comment by brianjgeiger** [comment]
django_rest_framework

**Comment by sloria** [comment]
That wasn't added in this PR; if there are issues with that, they should be addressed separately.

**Comment by brianjgeiger** [comment]
Okay.

**Comment by sloria** [comment]
I added the following

```
NodeSerializer
/data/attributes/fork
/data/relationships/forked_from

NodeRegistrationSerializer
/data/attributes/date_registered
/data/relationships/registered_from
```

**Comment by brianjgeiger** [comment]
It's fine. It's just that, with the more things added, it's more obviously wrong. There's an issue being made for it.

**Comment by brianjgeiger** [comment]
Looks good. Ready for CR2. :penguin:

File: website/project/model.py (hunk 2 above)
Comment (cwisecarver) [inline website/project/model.py]: Not too familiar with ODM but here's where I would change the way that DRF handles the lookup_field if we were using the Django ORM. https://github.com/tomchristie/django-rest-framework/blob/3.1.1/rest_framework/relations.py#L180-L189 I believe you could just override get_object and you could make lookup_field deal with 'something.something'

File: tests/api_tests/nodes/test_serializers.py (hunk 1 above)
Comment (cwisecarver) [inline tests/api_tests/nodes/test_serializers.py]: If there's any value, I've got an iPython notebook using the jsonschema package that validates json against the JSON-API schema. It wouldn't be hard to work that into the tests. It also actually gives you a trace on where the error is in the JSON. I know it's *shocking*.

File: api/nodes/serializers.py [hunk 3]
```diff
@@ -78,12 +79,19 @@ class NodeSerializer(JSONAPISerializer):
...
                                                   lookup_url_kwarg='node_id', meta={'count': 'get_pointers_count'}))
 
-    parent = JSONAPIHyperlinkedIdentityField(view_name='nodes:node-detail', lookup_field='parent_id', link_type='self',
+    parent = JSONAPIHyperlinkedIdentityField(view_name='nodes:node-detail', lookup_field='parent_id', link_type='related',
                                               lookup_url_kwarg='node_id')
 
     registrations = DevOnly(JSONAPIHyperlinkedIdentityField(view_name='nodes:node-registrations', lookup_field='pk', link_type='related',
                                                      lookup_url_kwarg='node_id', meta={'count': 'get_registration_count'}))
 
+    forked_from = JSONAPIHyperlinkedIdentityField(
```
Comment (cwisecarver) [inline api/nodes/serializers.py]: I assume that Sam's PR will fix this but I am getting invalid json back from empty relationships.
```json
      "forked_from": {
        "links": {
          "related": {
            "href": null,
            "meta": {}
          }
        }
      }
```

File: tests/api_tests/nodes/test_serializers.py (hunk 1 above)
Comment (sloria) [inline tests/api_tests/nodes/test_serializers.py]: Ha, nice. Yeah, validating against the JSON-API schema is definitely a good thing. 

These tests serve a different purpose: testing that the serialized representations of our stored objects have the correct values. For example, these tests were added to ensure that the `registration_from` URLs that I added were actually correct.

Comment (cwisecarver) [inline tests/api_tests/nodes/test_serializers.py]: Makes sense. I just figured the form is what the schema is looking at and the last thing that **should** be touching the form is the serializer. Runscope tests are great but if something is changing the format of the json between the serializer and the wire we've got problems.

File: api/nodes/serializers.py (hunk 3 above)
Comment (sloria) [inline api/nodes/serializers.py]: Yes, this is addressed in Sam's PR.

**Pull request merged**